netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -788,8 +788,27 @@ private void cancelStream(ChannelHandlerContext ctx, CancelServerStreamCommand c
       PerfMark.linkIn(cmd.getLink());
       // Notify the listener if we haven't already.
       cmd.stream().transportReportStatus(cmd.reason());
-      // Terminate the stream.
-      encoder().writeRstStream(ctx, cmd.stream().id(), Http2Error.CANCEL.code(), promise);
+
+      // Now we need to decide how we're going to notify the peer that this stream is closed.
+      // If possible, it's nice to inform the peer _why_ this stream was cancelled by sending
+      // a structured headers frame. It is not always going to be possible (e.g., if we have
+      // already sent headers). If we can't send structured headers, we will fall back to
+      // sending a bare RST_STREAM.
+      boolean shouldCloseWithHeaders =
+              cmd.wantsHeaders() && !connection().stream(cmd.stream().id()).isHeadersSent();
```

Comment:
`connection().stream(cmd.stream().id())` can return `null`. We can treat that as headers sent. (We could potentially do nothing instead of sending RST_STREAM, but let's not change that old code flow.)

Reply:
Good call, I extracted this logic into a static helper method and documented it a bit